scripts/metadata_enrichment/tob_bioheart_sequencing_dates.py

```diff
@@ -0,0 +1,234 @@
+    fluidx_to_assay_ids_bioheart: defaultdict,
+) -> defaultdict:
+    """
+    Append two defaultDict objects into a single dictionary.
+    Verifies that there are no intersecting tube ids
+
+    :return: When no intersection of keys
+    """
+    result = defaultdict(list)
+    dicts_to_merge = [fluidx_to_assay_ids_tob, fluidx_to_assay_ids_bioheart]
+    # First check to ensure that there are no overlapping keys
+    set_tob = set(fluidx_to_assay_ids_tob.keys())
+    set_bioheart = set(fluidx_to_assay_ids_bioheart.keys())
+
+    if len(set_tob.intersection(set_bioheart)) == 0:
+        # Append bioheart dict to tob dict
+        for d in dicts_to_merge:
+            for key, value in d.items():
+                result[key].append(value)
+        return result
+    else:  # noqa: RET505
+        print(
+            f'Error: these keys intersect: {set_tob.intersection(set_bioheart)}',
+        )
+        return defaultdict()
+
+
+# TODO: combine pd.apply() into a single expression
+def extract_excel():
+    """
+    Reads in metadata stored in excel spreadsheets. Appends these into
+    a single dataframe and performs required transformations to columns
+    to extract the desired data.
+
+    :return: Fluid tube (key) IDS mapped to sequencing dates (value)
+    :rtype: dict
+    """
+    tob_workbook_names = [
+        'scripts/metadata_enrichment/1K1K Sequencing Dates.xlsx',
+        'scripts/metadata_enrichment/BioHEART Sequencing Dates.xlsx',
+    ]
+    df_list = []
+
+    # Amalgamate data in all sheets listed in tob_sheet_names and bioheart_sheet_names
+    for workbook in tob_workbook_names:
+        df_sheet_index = pd.read_excel(workbook, sheet_name=None, header=0)
+        temp_df = pd.concat(df_sheet_index)
+        df_list.append(temp_df)
+
+    aggregated_df = pd.concat(df_list, ignore_index=True)
+
+    # Separate accession date and fluidX_id in Sample Identifier column
+    aggregated_df['accession_date'] = aggregated_df['Sample Identifier'].apply(
+        lambda x: x.split('_')[0],
+    )
+    aggregated_df['fluidX_id'] = aggregated_df['Sample Identifier'].apply(
+        lambda x: x.split('_')[1],
+    )
+
+    # Insert accession value into new dictionary fluidX_to_sequencing_date. Key is FluidX ID
+    return pd.Series(
+        aggregated_df.accession_date.values,
+        index=aggregated_df.fluidX_id,
+    ).to_dict()
+
+
+# TODO: combine api calls and
+async def upsert_sequencing_dates(
+    fluidx_to_assay_ids: defaultdict,
+    fluidx_to_sequencing_date: dict,
+):
+    # construct API update calls
+    # Iterate through the fluidX_to_assay_IDs dictionary because this is representative of what's already in metamist
+    # That is: fluidX_to_assay_IDs groups assays by fluidX_tube_id
+
+    assay_api = AssayApi()
+    assays_to_update = []
+
+    for fluidx_id, assay_ids in fluidx_to_assay_ids.items():
+        # Get sequencing date for each fluidX id from fluidX_to_sequencing_date dict
+        if fluidx_id:
```

Comment:
You can reduce a level of indentation here with a so-called "early return", which is using a continue to go to the next stage of the loop if a condition is not met. This removes the need for an if-else and knocks out one indentation of nesting. 

Something like:
```python
for fluidx_id, assay_ids in fluidx_to_assay_ids.items():
    if not fluidx_id:
        print(f'*****\nNO TUBE ID FOR : {fluidx_id} and assays {assay_ids}')
        continue
    try:
        ...

Reply:
Implemented. Great suggestion.